<b>Hola Adriana</b>

Mi nombre es Alejandro Abia y tengo el gusto de revisar tu proyecto.

A continuación, encontrarás mis comentarios en celdas pintadas de tres colores (verde, amarillo y rojo), a manera de semáforo. Por favor, <b>no las borres ni muevas de posición</b> mientras dure el proceso de revisión.

<div class="alert alert-block alert-success">
<b>Éxito</b> <a class="tocSkip"></a>
En celdas verdes encontrarás comentarios en relación a tus aciertos y fortalezas.
</div>
<div class="alert alert-block alert-warning">
<b>Antención</b> <a class="tocSkip"></a>
Utilizaré el color amarillo para llamar tu atención, expresar algo importante o compartirte alguna idea de valor.
</div>
<div class="alert alert-block alert-danger">
<b>A resolver</b> <a class="tocSkip"></a>
En rojo emitiré aquellos puntos que deberás atender para aprobar la revisión.
</div>
<div class="alert alert-block alert-info">
<b>Comentario estudiante</b><a class="tocSkip"></a>
Es factible que, a lo largo del proceso de revisión, quieras dejarme comentarios. Si es el caso, por favor realízalo dentro de celdas azules como esta.
</div>
Respecto del proceso de revisión, tu proyecto será aceptado una vez que los comentarios en rojo hayan sido atendidos.
¡Empecemos!

# SPRINT  11
En este ejercicio se plantea la prediccion de que region generar mayor ganancia para crear 200 nuevos pozos, se analizan las caracteristicas, se entrena un modelo de regresion lineal. Se hace el calculo financiero acerca de la inversion incial y cuanto es la ganancia por miles de barriles, para conocer el numero de barriles esperados para tener una ganacia o perdida. De igual forma se contempla cual region tiene mayor ganancia y se toma en cuanta la perdida. 

In [2]:
#importamos librerias que vamos a usar
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

In [4]:
#lectura de los datasets que usaremos
df_0=pd.read_csv('/geo_data_0.csv') # /Users/EQI/Documents/ATT/S11
df_1=pd.read_csv('/geo_data_1.csv')
df_2=pd.read_csv('/geo_data_2.csv')

In [6]:
#Funcion para observar los datos, info, mediaa, maxima, duplicados, nulos
def observar_datos(data, head=5):
    print("##################### info #####################")
    print()
    print(data.info())
    print("##################### sample #####################")
    print()
    print(data.sample(head))
    print("##################### describe #####################")
    print()
    print(data.describe())
    print("##################### duplicados #####################")
    print()
    print(data.duplicated().sum())
    print("##################### nulos #####################")
    print()
    print(data.isna().sum())

In [8]:
#aplicacion de funcion para prepar datos en el primer dataset
observar_datos(df_0)

##################### info #####################

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None
##################### sample #####################

          id        f0        f1        f2     product
26215  4woJ0  1.088629  0.267216  4.840734   96.220145
70120  dZwGG  0.082447  0.529859  3.861110   67.545915
72708  ooHXp  0.629340  0.771378  6.706266   94.249275
82840  88oIt  0.802635 -0.452491  2.741313  106.925514
28519  wPSWO  1.708366 -0.278757  2.812235  112.299561
##################### describe #####################

                  f0             f1             f2        product
count  10

In [8]:
#aplicacion de funcion para prepar datos en el segundo dataset
observar_datos(df_1)

##################### info #####################

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None
##################### sample #####################

          id         f0         f1        f2     product
34964  93KSq   7.020451   2.243472  1.995881   53.906522
79837  6EcEM  -6.886074 -10.436939  2.009555   57.085625
3574   jMnVc  -5.236074  -3.381795  4.995996  137.945408
31371  3GWgU  -3.734903  -5.148355  3.999647  110.992147
91785  xAE6E  16.413674  -0.093560  2.006959   53.906522
##################### describe #####################

                  f0             f1             f2        produ

In [10]:
#aplicacion de funcion para prepar datos en el tercer dataset
observar_datos(df_2)

##################### info #####################

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None
##################### sample #####################

          id        f0        f1        f2     product
92726  IKGBV -0.777947  2.299311  8.227096  107.775125
27757  eDave  0.692848  0.164918  6.751146  126.783282
71569  o4BCv  0.524936  0.337644  2.098238    3.248437
55126  zgn0U  0.096088 -0.322009  4.349893   83.540892
51876  pN2Qp  1.199831 -0.158462  3.724348   74.108057
##################### describe #####################

                  f0             f1             f2        product
count  10

<div class="alert alert-block alert-success">
<b>Celdas [2–10]</b> <a class="tocSkip"></a><br>
Excelente función para exploración exploratoria. La implementación de <code>observar_datos()</code> es práctica y completa, ya que revisa tipos de datos, valores extremos, duplicados y nulos. Además, aplicas la función de forma consistente sobre los tres datasets, lo que asegura una base de calidad para el análisis posterior.
</div>


Observamos que cada region contiene 100,000 pozos, los valores de f0,f1,f2 ya estan normalizados, y no se cuenta con valores nulos en los 3 datasets. 

id — identificador único de pozo de petróleo

product — volumen de reservas en el pozo de petróleo (miles de barriles).

f0, f1, f2 — tres características de los puntos

# **Modelo de regresion lineal para predicción de volumen de reserva de pozos**

In [14]:
#Debido a que es una prediccion cuantitativa se toma el modelo de regresion lineal
state=np.random.RandomState(12345)
def crear_modelo_entrenar_predecir(data, state):
  target = data['product']
  features = data.drop(['product', 'id'], axis=1)

  # segmentación en datos de entrenamiento, validacion y prueba
  features_train,features_valid, target_train, target_valid = train_test_split(features, target, test_size=0.25, random_state=state)
  model = LinearRegression() # inicializa el constructor de modelos
  model.fit(features_train, target_train) # entrena el modelo en el conjunto de entrenamiento
  predictions = model.predict(features_valid) # obtén las predicciones del modelo en el conjunto de validación

   # calcula la RECM en el conjunto de validación.
   #la función mean_squared_error calcula el error medio al cuadrado, pero para la unidad de medida de los barriles de petroleo es incoherente
   #expresarla al cuadrado, entonces calculamos la raíz cuadrada.
  rmse =np.sqrt(mean_squared_error(target_valid, predictions))

  #Muestra en la pantalla el volumen promedio de las reservas previstas y el RECM del modelo
  print("RMSE", rmse)

  print("Volumen de barriles promedio predecidos", predictions.mean())
  print("Volumen de barriles promedio reales", target_valid.mean())

  #Guarda las predicciones y las respuestas correctas para el conjunto de validación
  resultado=pd.DataFrame(data={'valor_real':target_valid, 'prediccion_valores':predictions,})
  return resultado

In [16]:
#Creamos el modelo de regresión lineal para cada REGION
regiones= ['Region 1', 'Region 2', 'Region 3']
datos=[df_0, df_1, df_2]
predict_region={} #diccionario para almacenar 3 datafreme que retornó la función crear_modelo_entrenar_predecir
for datos, region in zip(datos,regiones):
  print(region)
  predict_region[region]=crear_modelo_entrenar_predecir(datos, state)




Region 1
RMSE 37.5794217150813
Volumen de barriles promedio predecidos 92.59256778438035
Volumen de barriles promedio reales 92.07859674082927
Region 2
RMSE 0.8897367737680658
Volumen de barriles promedio predecidos 68.76995145799752
Volumen de barriles promedio reales 68.77162424984647
Region 3
RMSE 39.958042459521614
Volumen de barriles promedio predecidos 95.087528122523
Volumen de barriles promedio reales 94.74895871720241


<div class="alert alert-block alert-success">
<b>Celdas [14–16]</b> <a class="tocSkip"></a><br>
Muy bien planteada la función para entrenar y validar modelos por región. Separas adecuadamente variables objetivo y predictoras, manejas el split correctamente y usas RMSE como métrica relevante para el contexto. También es positivo que muestres tanto el valor medio predicho como el real, lo cual facilita la interpretación del modelo.
</div>


observamos que la región 3 es la que tiene el mayor volumen de barriles extraidos, que es de 94.74 y el menor error de prediccion lo tiene la region 2 con 0.88 miles de barriles.

In [19]:
#mostramos el diccionario que contiene 2 columnas datos reales y predicciones
predict_region

{'Region 1':        valor_real  prediccion_valores
 71751   10.038645           95.894952
 80493  114.551489           77.572583
 2655   132.603635           77.892640
 53233  169.072125           90.175134
 91141  122.325180           70.510088
 ...           ...                 ...
 12581  170.116726          103.037104
 18456   93.632175           85.403255
 73035  127.352259           61.509833
 63834   99.782700          118.180397
 43558  177.821022          118.169392
 
 [25000 rows x 2 columns],
 'Region 2':        valor_real  prediccion_valores
 70156   53.906522           54.238152
 42796   84.038886           83.217878
 36723    3.179103            2.971467
 79446   26.953261           25.916480
 93328  110.992147          111.341137
 ...           ...                 ...
 5864    26.953261           26.722422
 69826   80.859783           81.570895
 47036   84.038886           83.901769
 93329    0.000000           -1.259473
 93430   26.953261           26.169570
 
 [25000 r

# **cálculo de ganancias  **

In [21]:
presupuesto=100000000   #cien millones de dolares presupuesto para 200 pozos, inversion inicial
incomes=4500     #Se genera 4,500 dolares por cada 1000 barriles extraídos
pozos=200  #nuevos que se desean explotar


In [29]:
# cuanto es el minimo de miles de barriles que necesito para reponer la inversion inicial 
volumen_minimo=presupuesto/incomes
print(volumen_minimo, 'miles de barriles en total para recuperar lo invertido')

22222.222222222223 miles de barriles en total para recuperar lo invertido


In [33]:
#calcula el volumen minimo que debe tener cada pozo para llegar al punto de equilibrio
volumen_minimo_pozo=volumen_minimo/pozos
print(volumen_minimo_pozo, 'miles de barriles debe generar cada pozo')

111.11111111111111 miles de barriles debe generar cada pozo


<div class="alert alert-block alert-success">
<b>Celdas [21, 29, 33]</b> <a class="tocSkip"></a><br>
Buena lógica para calcular el punto de equilibrio. Derivas correctamente la cantidad mínima de barriles necesarios para recuperar la inversión, tanto a nivel global como por pozo. Esto proporciona un contexto financiero claro antes de evaluar las ganancias proyectadas.
</div>


In [39]:
#Dentro de una región ubicamos a los 200 mejores pozos, para ello los ordenamos en forma descendente, hacemos la suma del volumen de
#miles de barriles generados y sacamos el margen de ganancia al miltiplocar el ingreso por todos los miles de barriles y restar la inversion
def generacion(datos):
  mejores_lugares=datos.sort_values(by='prediccion_valores', ascending=False)['valor_real'].head(pozos)
  volumen_mejores_lugares=mejores_lugares.sum()
  margen=((volumen_mejores_lugares*incomes)-presupuesto)
  return margen

<div class="alert alert-block alert-success">
<b>Celda [39]</b> <a class="tocSkip"></a><br>
La función <code>generacion()</code> está bien diseñada: seleccionas los mejores pozos con base en la predicción, pero usas los valores reales para calcular las ganancias, lo cual es un excelente enfoque para evitar fuga de datos. Bien pensado.
</div>


In [41]:
generacion(predict_region['Region 1'])

33208260.43139851

In [43]:
generacion(predict_region['Region 2'])

24150866.966815114

In [45]:
generacion(predict_region['Region 3'])

25399159.45842947

In [47]:
#Aplicamos herramienta de bootstraping para calcular la ganancia de un conjunto de pozos de petróleo seleccionados 
#Esta función toma al azar un conjunto de 500 ubicaciones dentro de una región y
#a ese conjunto se le calcula el beneficio que finalmente se agrega a la lista beneficio_muestra
# y ésto se repite 1000 veces

def modelo_bootstrap(datos, n_iteraciones=1000, state=state):
  generacion_muestra=[]
  for i in range(n_iteraciones):
    ubicaciones=datos.sample(n=500, replace=True, random_state=state)
    generacion_muestra.append(generacion(ubicaciones))
  generacion_muestra=pd.Series(generacion_muestra)
  return generacion_muestra

In [49]:
modelo_bootstrap(predict_region['Region 1'])

0      3.345078e+06
1      4.730910e+06
2      4.383718e+06
3      3.185568e+06
4      3.688245e+06
           ...     
995    3.838511e+06
996    2.881459e+06
997    3.884321e+06
998    4.042372e+06
999    5.721960e+06
Length: 1000, dtype: float64

<div class="alert alert-block alert-warning">
<b>Celdas [47–49]</b> <a class="tocSkip"></a><br>
El uso de bootstrap es adecuado, pero hay un pequeño problema técnico: al usar <code>random_state=state</code> dentro del ciclo, la semilla se reinicia en cada iteración, generando muestras idénticas. Para evitarlo, remueve <code>random_state</code> de <code>sample()</code> dentro del bucle.
</div>


In [51]:
modelo_bootstrap(predict_region['Region 2'])

0      3.343157e+06
1      6.032541e+05
2      1.085565e+07
3      6.054448e+06
4      7.017923e+06
           ...     
995    5.162502e+06
996    4.463376e+06
997    6.310711e+06
998    7.930394e+06
999    3.999365e+06
Length: 1000, dtype: float64

In [53]:
modelo_bootstrap(predict_region['Region 3'])

0      1.551682e+05
1      2.383618e+06
2      6.365848e+06
3      2.526721e+06
4      3.418216e+06
           ...     
995    4.473636e+06
996   -4.395696e+05
997    6.210045e+06
998    2.101507e+06
999    6.710192e+06
Length: 1000, dtype: float64

In [55]:
#Intervalo de confianza para concer la ganancia o perdida promedio
def metricas(generacion_muestra, region):
  nivel_confianza=0.95
  alpha=1-nivel_confianza
  generacion_promedio=generacion_muestra.mean()
  perdida=(generacion_muestra[generacion_muestra<0]).mean()
  print("____________________")
  print("Las metricas de la Region", region)
  print("Ganancia promedio", generacion_promedio)
  print("Pérdida promedio: ", perdida)

In [57]:
for region, data in predict_region.items():
  margen_region=modelo_bootstrap(data)
  metricas(margen_region, region)

____________________
Las metricas de la Region Region 1
Ganancia promedio 3931490.9009217
Pérdida promedio:  -1136092.9078568465
____________________
Las metricas de la Region Region 2
Ganancia promedio 4581817.322551863
Pérdida promedio:  -414086.3379975736
____________________
Las metricas de la Region Region 3
Ganancia promedio 3447468.317867259
Pérdida promedio:  -1118072.1634442045


<div class="alert alert-block alert-success">
<b>Celdas [55–57]</b> <a class="tocSkip"></a><br>
Buena presentación de métricas por región. El cálculo de pérdida promedio y ganancia promedio es útil para evaluar riesgo-retorno. Además, muestras claramente que la región 2 es la más rentable con el menor riesgo. Bien fundamentada tu elección final.
</div>


Conclusiones: AL aplicar al modelo bootstraping podemos generar mas muestras. La región 2 tiene la ganancia más alta de 4,501,150 y el riesgo de pérdida es la más baja de las tres regiones con una pérdida de 414,086


<div class="alert alert-block alert-success">
<b>Comentario final</b> <a class="tocSkip"></a><br>
¡Excelente trabajo, Adriana! Tu proyecto está bien estructurado y refleja una comprensión clara tanto del modelado predictivo como de su aplicación financiera. Usaste correctamente regresión lineal, métricas relevantes, validación cruzada y bootstrap para evaluar el riesgo.  
<br><br>
🔍 Como sugerencia para tu versión final de portafolio: considera corregir el uso de <code>random_state</code> en el bootstrap para asegurar muestras aleatorias verdaderas. También sería útil matizar tus conclusiones sobre cuál región es mejor en función no solo del volumen promedio, sino también del error y la estabilidad del modelo.  
<br><br>
¡Sigue trabajando con este nivel de detalle y reflexión, vas por muy buen camino!
</div>
